raft.go

```diff
@@ -1478,10 +1478,13 @@ func stepLeader(r *raft, m pb.Message) error {
 				switch {
 				case pr.State == tracker.StateProbe:
 					pr.BecomeReplicate()
-				case pr.State == tracker.StateSnapshot && pr.Match >= pr.PendingSnapshot:
-					// TODO(tbg): we should also enter this branch if a snapshot is
-					// received that is below pr.PendingSnapshot but which makes it
-					// possible to use the log again.
+				case pr.State == tracker.StateSnapshot && pr.Match+1 >= r.raftLog.firstIndex():
```

Comment:
Logically, it can also be (maybe should?) `pr.Next >= r.raftLog.firstIndex()`.

`Next` is the next index to send. Indices between `Match+1` and `Next-1` are in-flight (by invariant which we haven't formalized yet), so don't necessarily need to be present in the log.

The `MaybeUpdate()` [call](https://github.com/etcd-io/raft/blob/25e4bb1f4ed7fa19a9dea4a8b5f8cea13d048bc6/tracker/progress.go#L173) 4 lines above should set `Next` to be at least `Match+1`, so this should be safe.

Feel free to leave this for #124.

Reply:
Let's do it for #124. Semantically, `Match+1` seems better since that's where the leader's and follower's logs match, but they're equivalent here and I don't have a particularly strong opinion either way.